core/core-configuration-layer.el

```diff
@@ -155,60 +143,55 @@ NAME."
         (replace-match name t)))
     (save-buffer)))
 
-(defun configuration-layer//get-contrib-category-dirs ()
-  "Return a list of all absolute paths to the contribution categories stored
-in `configuration-layer-contrib-categories'"
-  (mapcar
-   (lambda (d) (expand-file-name
-                (concat configuration-layer-contrib-directory (format "%s/" d))))
-   configuration-layer-contrib-categories))
+(defun configuration-layer//is-layer-p (path)
+  "Return non-nil if the path given is a valid configuration layer."
+  (when (file-directory-p path)
+    (let ((files (directory-files path)))
+      (or (member "packages.el" files)
+          (member "extensions.el" files)))))
 
 (defun configuration-layer//discover-layers ()
   "Return a hash table where the key is the layer symbol and the value is its
 path."
-  (let ((contrib-cat-dirs (configuration-layer//get-contrib-category-dirs))
+  ;; load private layers at the end on purpose
+  ;; we asume that the user layers must have the final word
+  ;; on configuration choices.
+  (let ((search-paths (append (list configuration-layer-contrib-directory)
+                              dotspacemacs-configuration-layer-path
+                              (list configuration-layer-private-directory)))
         (discovered '())
         (result (make-hash-table :size 256)))
-    (setq discovered
-          (append discovered (configuration-layer//discover-layers-in-dir
-                              configuration-layer-contrib-directory
-                              configuration-layer-contrib-categories)))
-    (dolist (dir (append contrib-cat-dirs
-                         dotspacemacs-configuration-layer-path))
-      (setq discovered
-            (append discovered (configuration-layer//discover-layers-in-dir
-                                dir))))
-    ;; load private layers at the end on purpose
-    ;; we asume that the user layers must have the final word
-    ;; on configuration choices.
-    (setq discovered
-          (append discovered (configuration-layer//discover-layers-in-dir
-                              configuration-layer-private-directory
-                              '("snippets"))))
-    ;; add spacemacs layer
+    ;; depth-first search of subdirectories
```

Comment:
Does this affect startup time significantly? It probably doesn't but I can't say that for certain.